### DISCUSSION TITLE
[Collaboration and Question] About the updated Firebase iOS Bindings

### DISCUSSION THREAD

**Discussion by vhugogarcia** [AdamEssenmacher/GoogleApisForiOSComponents#6] [General]
Hello @AdamEssenmacher , first, I would like to thank you for your effort to help us with the Firebase Bindings, really appreciate it.

I would like also to collaborate by updating the Plugin.Firebase from our friend Tobias, in fact I went ahead and updated the projects to use the latest version from your packages. You can find the updates here: https://github.com/vhugogarcia/Plugin.Firebase/tree/feature/packages-ios-upgrade

However, when trying to run a sample app to run some tests, I found that when building the iOS project, it returns the following error and warnings:
<img width="1418" alt="image" src="https://github.com/AdamEssenmacher/GoogleApisForiOSComponents/assets/1047398/9d9cdc69-ac4a-4e40-a4a9-2f971f87a915">

I have set the mtouch arguments as you mentioned it on the README.
```xml
<PropertyGroup Condition="'$(Configuration)|$(TargetFramework)|$(Platform)'=='Debug|net8.0-ios|AnyCPU'">
	  <MtouchLink>None</MtouchLink>
	  <CreatePackage>false</CreatePackage>
	  <MtouchExtraArgs>--gcc_flags -L/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/iphonesimulator/</MtouchExtraArgs>
	</PropertyGroup>

	<PropertyGroup Condition="'$(Configuration)|$(TargetFramework)|$(Platform)'=='Release|net8.0-ios|AnyCPU'">
	  <CreatePackage>false</CreatePackage>
	  <RuntimeIdentifier>ios-arm64</RuntimeIdentifier>
	  <CodesignProvision>App Distribution</CodesignProvision>
	  <CodesignKey>iPhone Distribution</CodesignKey>
	  <CodesignEntitlements>Platforms\iOS\Entitlements.plist</CodesignEntitlements>
	  <MtouchLink>SdkOnly</MtouchLink>
	  <MtouchUseLlvm>true</MtouchUseLlvm>
	  <MtouchExtraArgs>--gcc_flags -L/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/iphoneos</MtouchExtraArgs>
	</PropertyGroup>
```

do you know if I'm missing something on my side? 

Thanks in advance

**Comment by AdamEssenmacher** [comment]
Wow, thanks for being proactive!

I'm a collaborator over at Plugin.Firebase, and Tobias and I already have this planned. I'm working on it now.

That said, I'm focusing on the features that I personally rely on, as I can validate the updates naturally without going too far out of my way. At the moment, notable exclusions are:

- Auth.Facebook
- Auth.Google
- Crashlytics
- DynamicLinks
- RemoteConfig

If you're using any of these, I could use help testing them out.

I can also use help validating / testing these packages generally too :-)

**Comment by vhugogarcia** [comment]
Thanks @AdamEssenmacher for the response! Sounds like a plan! I also have apps using the bindings you upgraded only. 

However, I will be more than happy to help you test them on real applications. Just let me know how I can collaborate with. 

I will continue keeping an eye on your repository for new updates and releases. If there is a private branch you want me to jump on and run some tests, I will be happy to do it. 

Thanks again for the effort and looking forward for the new releases, this is going to be awesome 👏🏻

**Comment by AdamEssenmacher** [comment]
Those linker errors are addressed in the readme now.

**Reply by vhugogarcia** [reply]
Thanks @AdamEssenmacher let me look into.

**Reply by vhugogarcia** [reply]
@AdamEssenmacher thanks again for updating the README with the notes. I already updated my sample MAUI app's project with the Target for the swift links and the previous error is gone. However, I'm now facing this issue when building the project:

```
/usr/local/share/dotnet/packs/Microsoft.iOS.Sdk/17.2.8043/targets/Xamarin.Shared.Sdk.targets(3,3): Error: clang++ exited with code 1:
ld: warning: Could not find or use auto-linked framework 'GoogleAppMeasurement'
Undefined symbols for architecture arm64:
          "_APMAnalyticsConfiguration", referenced from:
              +[FIRAnalytics startWithConfiguration:options:] in FirebaseAnalytics(FIRAnalytics.o)
          "_APMAppMeasurementOriginFirebase", referenced from:
              +[FIRAnalytics startWithConfiguration:options:] in FirebaseAnalytics(FIRAnalytics.o)
          "_APMConsentSettings3P", referenced from:
              +[FIRAnalytics setConsent:] in FirebaseAnalytics(FIRAnalytics.o)
          "_APMFormattedEventName", referenced from:
              +[FIRAnalytics logEventWithOrigin:name:parameters:] in FirebaseAnalytics(FIRAnalytics.o)
          "_APMFormattedUserPropertyName", referenced from:
              +[FIRAnalytics setUserPropertyString:forName:] in FirebaseAnalytics(FIRAnalytics.o)
          "_APMIsAnalyticsCollectionDeactivated", referenced from:
              +[FIRAnalytics startWithConfiguration:options:] in FirebaseAnalytics(FIRAnalytics.o)
          "_APMIsAnalyticsCollectionEnabled", referenced from:
              +[FIRAnalytics startWithConfiguration:options:] in FirebaseAnalytics(FIRAnalytics.o)
          "_APMMonitorLogTagOptionKey", referenced from:
              +[FIRAnalytics startWithConfiguration:options:] in FirebaseAnalytics(FIRAnalytics.o)
          "_APMUserDataFieldEmailAddress", referenced from:
              +[FIRAnalytics initiateOnDeviceConversionMeasurementWithEmailAddress:] in FirebaseAnalytics(FIRAnalytics.o)
          "_APMUserDataFieldPhoneNumber", referenced from:
              +[FIRAnalytics initiateOnDeviceConversionMeasurementWithPhoneNumber:] in FirebaseAnalytics(FIRAnalytics.o)
          "_OBJC_CLASS_$_APMAdExposureReporter", referenced from:
              _OBJC_CLASS_$_FIRAAdExposureReporter in FirebaseAnalytics(FIRAAdExposureReporter.o)
          "_OBJC_CLASS_$_APMAnalytics", referenced from:
              objc-class-ref in FirebaseAnalytics(FIRAnalytics.o)
          "_OBJC_CLASS_$_APMConditionalUserProperty", referenced from:
              _OBJC_CLASS_$_FIRAConditionalUserProperty in FirebaseAnalytics(FIRAConditionalUserProperty.o)
          "_OBJC_CLASS_$_APMConditionalUserPropertyController", referenced from:
              _OBJC_CLASS_$_FIRAConditionalUserPropertyController in FirebaseAnalytics(FIRAConditionalUserPropertyController.o)
          "_OBJC_CLASS_$_APMEvent", referenced from:
              _OBJC_CLASS_$_FIRAEvent in FirebaseAnalytics(FIRAEvent.o)
          "_OBJC_CLASS_$_APMIdentifiers", referenced from:
              _OBJC_CLASS_$_FIRAIdentifiers in FirebaseAnalytics(FIRAIdentifiers.o)
          "_OBJC_CLASS_$_APMIdentity", referenced from:
              objc-class-ref in FirebaseAnalytics(FIRAnalytics.o)
          "_OBJC_CLASS_$_APMMeasurement", referenced from:
              objc-class-ref in FirebaseAnalytics(FIRAnalytics.o)
              objc-class-ref in FirebaseAnalytics(FIRAMeasurement.o)
              _OBJC_CLASS_$_FIRAMeasurement in FirebaseAnalytics(FIRAMeasurement.o)
          "_OBJC_CLASS_$_APMScreenViewReporter", referenced from:
              objc-class-ref in FirebaseAnalytics(FIRAScreenViewReporter.o)
              _OBJC_CLASS_$_FIRAScreenViewReporter in FirebaseAnalytics(FIRAScreenViewReporter.o)
          "_OBJC_CLASS_$_APMUserAttribute", referenced from:
              _OBJC_CLASS_$_FIRAUserAttribute in FirebaseAnalytics(FIRAUserAttribute.o)
          "_OBJC_CLASS_$_APMValue", referenced from:
              _OBJC_CLASS_$_FIRAValue in FirebaseAnalytics(FIRAValue.o)
          "_OBJC_METACLASS_$_APMAdExposureReporter", referenced from:
              _OBJC_METACLASS_$_FIRAAdExposureReporter in FirebaseAnalytics(FIRAAdExposureReporter.o)
          "_OBJC_METACLASS_$_APMConditionalUserProperty", referenced from:
              _OBJC_METACLASS_$_FIRAConditionalUserProperty in FirebaseAnalytics(FIRAConditionalUserProperty.o)
          "_OBJC_METACLASS_$_APMConditionalUserPropertyController", referenced from:
              _OBJC_METACLASS_$_FIRAConditionalUserPropertyController in FirebaseAnalytics(FIRAConditionalUserPropertyController.o)
          "_OBJC_METACLASS_$_APMEvent", referenced from:
              _OBJC_METACLASS_$_FIRAEvent in FirebaseAnalytics(FIRAEvent.o)
          "_OBJC_METACLASS_$_APMIdentifiers", referenced from:
              _OBJC_METACLASS_$_FIRAIdentifiers in FirebaseAnalytics(FIRAIdentifiers.o)
          "_OBJC_METACLASS_$_APMMeasurement", referenced from:
              _OBJC_METACLASS_$_FIRAMeasurement in FirebaseAnalytics(FIRAMeasurement.o)
          "_OBJC_METACLASS_$_APMScreenViewReporter", referenced from:
              _OBJC_METACLASS_$_FIRAScreenViewReporter in FirebaseAnalytics(FIRAScreenViewReporter.o)
          "_OBJC_METACLASS_$_APMUserAttribute", referenced from:
              _OBJC_METACLASS_$_FIRAUserAttribute in FirebaseAnalytics(FIRAUserAttribute.o)
          "_OBJC_METACLASS_$_APMValue", referenced from:
              _OBJC_METACLASS_$_FIRAValue in FirebaseAnalytics(FIRAValue.o)
```

Also I set the mtouchargs for iOS on my app's sample project:
`<MtouchExtraArgs>--gcc_flags -L/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift/iphoneos</MtouchExtraArgs>`

I'm trying to build it from macOS M1 Pro, which uses ARM architecture, I'm not sure if this helps or no.

**Reply by AdamEssenmacher** [reply]
The -gcc_flags should only be needed for Xamarin, not net6.

The original screenshot you posted showed that the GAppM-10.17.0 directory doesn't exist in your XamarinBuildDownload folder. It's possible that cache is bad.

Are you using Visual Studio on Windows by chance?

**Reply by vhugogarcia** [reply]
uh oh! I got confuse, I was thinking it should be added as well. I'm on macOS Sonoma right now. let me remove the mtouchargs and include Xamarin.Build.Download 0.11.4 and try again.

**Reply by AdamEssenmacher** [reply]
I just saw an issue reporting similar build warnings if you try to mix and match AdamE.Firebase.* packages with Xamarin.Firebase.* packages. That will never work.

**Reply by vhugogarcia** [reply]
thanks @AdamEssenmacher for getting back to me. I am unable to compile the sample project, I keep getting the same error shown above. I even included Xamarin.Build.Download 0.11.4 on the sample project without luck.

I found some people also faced similar issues with the missing architecture, I'm wondering if the bindings are right now being generated only for x86-64 and maybe we need to update the setting to include also ARM, I'm just thinking loud. I will continue debugging and see if I can find anything.

https://stackoverflow.com/questions/72344149/how-to-solve-xcode-build-error-undefined-symbols-for-architecture-x86-64

btw, I'm not mixing AdamE.Firebase packages with Xamarin.Firebase packages. The Sample project is pretty clean on that regard, hehe 😃 . 
<img width="968" alt="image" src="https://github.com/AdamEssenmacher/GoogleApisForiOSComponents/assets/1047398/a18fe904-3976-4b00-9c8d-5cde0373fa37">

**Reply by AdamEssenmacher** [reply]
The bindings are being built for multiple architectures. You can dig into the xcframeworks and check for yourself.

I have to ask this again, because the screenshots you're posting look very Visual Studio-y: Are you using Visual Studio on Windows by chance?

**Reply by vhugogarcia** [reply]
> The bindings are being built for multiple architectures. You can dig into the xcframeworks and check for yourself.
> 
> 
> 
> I have to ask this again, because the screenshots you're posting look very Visual Studio-y: Are you using Visual Studio on Windows by chance?

I am using Visual Studio for Mac, I also tried Visual Studio Code on macOS. This is very strange

**Reply by vhugogarcia** [reply]
@AdamEssenmacher I have some news, I updated to the latest version 10.19.0 of the packages in all projects, and also just for testing purposes, I removed the Firebase Analytics reference to see if I can get rid off the errors with the analytics and they are gone, however, I keep getting this error now:

`/Users/demo/.nuget/packages/adame.firebase.ios.core/10.19.0/buildTransitive/AdamE.Firebase.iOS.Core.targets(5,5): Error MSB3073: The command "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/bitcode_strip  -r -o " exited with code 1. (MSB3073)`

I have installed XCode latest version in macOS on the default path on my Mac, however, I'm not clear on what the error is about. After some researching, I believe the error is due to the target with the swift we added into the csproj file. I'm still unsure. I asked a few questions on Discord's DotNetEvolution server > #Apple channel, to see if somebody has any insights about the first error and then I will ask about this. This is my sample .csproj file: https://gist.github.com/vhugogarcia/c62e7999251014d725f544fe3a64d849

Just reporting here the progress, I will continue learning and digging into. 

Thank you 😃

**Comment by vhugogarcia** [comment]
Hello @AdamEssenmacher , I have **fantastic** news 🥳 !

I updated the packages to version 10.20.0 on the Plugin.Firebase fork I created, and I was able to build without errors the sample MAUI project, I also ran it on a physical device (debug) mode using a personal project, and they worked very good. https://github.com/vhugogarcia/Plugin.Firebase/tree/feature/packages-ios-upgrade

What I have tested so far and confirmed it worked as expected:

- Firebase Analytics
- Firebase Auth (Email authentication)
- Firebase Firestore
- Firebase Functions
- Firebase CloudMessaging

I plan to test the Storage, Crashlytics and the others as well later on. 
Note: The errors I experienced on version 10.18 and 10.19 are gone. There are a few warnings about some platform API deprecation, but that is something we can improve later on as well. So, far this is looking AMAZING!

Thanks @AdamEssenmacher for the contribution on this!

**Reply by AdamEssenmacher** [reply]
I didn't change anything on 10.18/10.19, so things suddenly working would point to a cache issue or similar. Unless you added a dependency to Installations; I think there might be something going on there...

**Reply by vhugogarcia** [reply]
I did not change anything, it is very weird! When I put back the version 10.19.0 I get errors again, maybe a cache issue as you mentioned, however, I tried doing a NuGet cleanup without luck, but using 10.20.0 all is good. Very weird! The good thing is that it is building and running as expected now in the latest version. I will continue running tests with the other libraries you pushed as alpha versions 😎👍🏻 and report the results back here. Thanks @AdamEssenmacher

**Comment by vhugogarcia** [comment]
Hello @AdamEssenmacher , I updated the Plugin.Firebase fork to use the AdamE iOS Firebase packages **10.22.0** and it worked very good for the following ones 🎊 :

- Firebase Analytics
- Firebase Auth _(Email authentication)_
- Firebase Firestore
- Firebase Functions
- Firebase CloudMessaging
- Firebase Storage

However, I started to test **Crashlytics** on the sample app v10.22.0, when I'm building the project, I get the following error:
`/Users/demo/.nuget/packages/xamarin.build.download/0.11.4/buildTransitive/Xamarin.Build.Download.targets(52,3): error XBD020: Invalid item ID FCrshlytcs-10.22ß.0`

I added the property group as mentioned on the README and added the reference to Installations latest version:
```xml
<PropertyGroup Condition="'$(TargetFramework)' == 'net8.0-ios'">
       <_ExportSymbolsExplicitly>false</_ExportSymbolsExplicitly>
</PropertyGroup> 
```

I'm not sure if I'm missing something else to add to the _csproj_, or if there is a typo somewhere, or if I am doing it wrong.

Thanks in advance!

**Reply by AdamEssenmacher** [reply]
Oops. Expect Crashlytics to be fixed in the 10.23+ iterations.

**Reply by vhugogarcia** [reply]
Thanks @AdamEssenmacher , I updated the packages to **10.23.0**, and I can confirm the Crashlytics building issue has been fixed. I tested it on a sample app using a real device and I was able to log a non-fatal error:

<img width="978" alt="image" src="https://github.com/AdamEssenmacher/GoogleApisForiOSComponents/assets/1047398/3b8bca6c-1d7d-4d9e-9dc8-8b76c93ce5d2">

Also, updated the fork I created from Plugin.Firebase on this branch with the latest versions: https://github.com/vhugogarcia/Plugin.Firebase/tree/feature/packages-ios-upgrade

Thanks Adam! You are basically at one last version of the Firebase SDK for iOS 😃

**Comment by moljac** [comment]
My iOS bindings knowledge got a bit rusty in last few years, but if you need any kind of help from Android side I am the one guilty of maintaining AX and GPS-FB-MLKit repos.

Thanks fellas for your work.

**Reply by AdamEssenmacher** [reply]
@moljac you're the best!

We'll be very happy if you just keep doing what you're doing 😊

**Reply by vhugogarcia** [reply]
@moljac you are the best! I also went ahead and updated the **Android SDKs** you have been working on, for the fork I have been collaborating on for the repository Plugin.Firebase: https://github.com/vhugogarcia/Plugin.Firebase/tree/feature/packages-ios-upgrade just for testing purposes and they worked very good. I know it is not an full test, since I'm only testing the services below for now, but I hope it helps other developers:

- Firebase Analytics
- Firebase Auth (Email authentication)
- Firebase Firestore
- Firebase Functions
- Firebase CloudMessaging
- Firebase Storage 
- Firebase Crashlytics

As Adam said, please continue giving us a hand to keep the Firebase SDKs for Android updated. THANK YOU SO MUCH!

**Reply by moljac** [reply]
Not the best, but not too bad.

I have to keep working on Android primarily, but I think/hope I will get more time in the future, so I can help with plugins and Android side and if I get even more spare time than with iOS and especially common API part.

**Reply by moljac** [reply]
>  I also went ahead and updated the Android SDKs you have been working on

Which one is that?

> please continue giving us a hand to keep the Firebase SDKs for Android updated.

I will. I promise.

**Reply by vhugogarcia** [reply]
I meant the Firebase Android SDKs references on the Plugin.Firebase fork I have been using for testing for example here: https://github.com/vhugogarcia/Plugin.Firebase/blob/feature/packages-ios-upgrade/src/Analytics/Analytics.csproj

```xml
<ItemGroup Condition="'$(TargetFramework)' == 'net8.0-android'">
        <PackageReference Include="Xamarin.Firebase.Analytics" Version="121.3.0.5" />
</ItemGroup>
```

I upgraded to the latest versions and all worked good for Android 👍🏻 😃

**Comment by vhugogarcia** [comment]
Hello @AdamEssenmacher , I updated the [Plugin.Firebase fork](https://github.com/vhugogarcia/Plugin.Firebase/tree/feature/packages-ios-upgrade) to use the AdamE iOS Firebase packages **10.24.0** and it worked very good for the following ones 😃 

- Firebase Analytics
- Firebase Auth (Email authentication)
- Firebase Firestore
- Firebase Functions
- Firebase CloudMessaging
- Firebase Storage
- Firebase Crashlytics

**NO** building issues and it runs OK on physical devices on DEBUG and RELEASE mode. This time it took me a little more testing since I was testing the new bindings on real devices on release mode, for example on Apple I released an app on TestFlight for testing and it worked OK, and the Apple Privacy Manifest warnings are gone. On Android, I published an internal version of a personal app confirming the app runs as expected on release mode, and the results were very good. 🥳 

Just out of curiosity, are you planing to upgrade the packages to **10.25.0**? https://firebase.google.com/support/release-notes/ios
